File: utils/common/test_trie.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.nicknames = set()

class NicknameSearch:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, nickname):
        node = self.root
        for char in nickname:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
            node.nicknames.add(nickname)

    def search(self, query):
        node = self.root
        for char in query:
            if char not in node.children:
                return []
            node = node.children[char]

        return list(node.nicknames)

# Example usage:
nickname_search = NicknameSearch()
nicknames = ["john", "jane", "jack", "jill", "hello world", "hi world"]
for nickname in nicknames:
    nickname_search.insert(nickname)

search_result = nickname_search.search("world")
#search_result = nickname_search.search("h")
print(search_result)  # Output: ['hello world']

File: utils/common/test_trie.cpp
#include <iostream>
#include <locale>
#include <unordered_map>

using std::unordered_map;
using std::string;
struct TrieNode {
    unordered_map<char, TrieNode*> children;
    bool isEndOfWord;
    TrieNode(): isEndOfWord(false) {}
};

class Trie {
private:
    TrieNode *root;
public:
    Trie() {
        root = new TrieNode();
    }
    
    void insert(const string& word) {
        TrieNode *currentNode = root;
        for (char ch: word) {
            if (currentNode->children.find(ch) == currentNode->children.end()) {
                currentNode->children[ch] = new TrieNode();
            }
            currentNode = currentNode->children[ch];
        }
        currentNode->isEndOfWord = true;
    }

    bool search(const string& word) {
        TrieNode *currentNode = root;
        for (char ch: word) {
            if (currentNode->children.find(ch) == currentNode->children.end()) {
                return false;
            }
            currentNode = currentNode->children[ch];
        }
        return currentNode->isEndOfWord;
    }

    bool startsWith(const string& prefix) {
        TrieNode *currentNode = root;
        for (char ch: prefix) {
            if (currentNode->children.find(ch) == currentNode->children.end()) {
                return false;
            }
            currentNode = currentNode->children[ch];
        }
        return true;
    }
};

int main() {
    Trie* trie = new Trie();

    trie->insert("hello");
    trie->insert("world");
    trie->insert("hi");

    std::cout << "Search 'hello': " << trie->search("hello") << std::endl; // true
    std::cout << "Search 'world': " << trie->search("world") << std::endl; // true
    std::cout << "Search 'hell': " << trie->search("hell") << std::endl;   // false
    std::cout << "Starts with 'wor': " << trie->startsWith("wor") << std::endl; // true
    std::cout << "Starts with 'woe': " << trie->startsWith("woe") << std::endl; // false

    return 0;
}

